File: PGS/Lv.0/안전지대.cpp
#include <string>
#include <vector>

using namespace std;

int N;
vector<vector<int>> map;
int dX[8] = { -1, -1, -1, 0, 0, 1, 1, 1 };
int dY[8] = { -1, 0, 1, -1, 1, -1, 0, 1 };

int isOut(int x, int y) {
	return (x < 0 || x >= N || y < 0 || y >= N);
}

void unsafe(int x, int y) {
	for (int i = 0; i < 8; i++) {
		int nX = x + dX[i];
		int nY = y + dY[i];

		if (isOut(nX, nY) || map[nX][nY] == 1) {
			continue;
		}

		map[nX][nY] = 1;
	}
}

int solution(vector<vector<int>> board) {
	int answer = 0;

	N = board.size();
	map.insert(map.begin(), board.begin(), board.end());

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			if (board[i][j] == 1) {
				unsafe(i, j);
			}
		}
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			if (map[i][j] == 0) {
				answer++;
			}
		}
	}

	return answer;
}

File: PGS/Lv.0/JAVA/안전지대.java
class Solution {
    
    static final int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};
    static final int[] dy = { -1, 0, 1, -1, 1, -1, 0, 1};
    static int n;
    static int[][] unsafe;
    
    boolean isOut(int x, int  y) {
        return (x < 0 || x >= n || y < 0 || y >= n);
    }
    
    void check(int[][] board, int x, int y) {
        for(int i = 0; i < 8; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(!isOut(nx, ny) && board[nx][ny] == 0) {
                unsafe[nx][ny] = 1;
            }
        }
    }
    
    public int solution(int[][] board) {
        
        n = board.length;
        unsafe = new int[n][n];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                unsafe[i][j] = board[i][j];
            }
        }
 
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(board[i][j] == 1) {
                    check(board, i, j);
                }
            }
        }
        
        int answer = n * n;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(unsafe[i][j] == 1) {
                    answer--;
                }
            }
        }
   
        return answer;
    }
}